README.md
# NetBox Virtual Circuit Plugin

A plugin for [NetBox](https://github.com/netbox-community/netbox) that supports
Virtual Circuit management.

## Definitions

A Virtual Circuit is a circuit created by routing two or more VLANs together.

Each Virtual Circuit is identified by a name, a numeric ID (0-32767), along with
a context. Each Virtual Circuit must be assigned one of the following operational
statuses:
- Pending Configuration
- Configured
- Pending Deletion
- Configuration Error

When a VLAN is assigned to a Virtual Circuit, it can not exist in another Virtual
Circuit without first being removed.

## Installing

Since the plugin is published on
[PyPI](https://pypi.org/project/netbox-virtual-circuit-plugin/), simply issue:
```python
pip install netbox-virtual-circuit-plugin
```
to download and install it.

To enable to plugin, add the plugin's name to the `PLUGINS` list in
`configuration.py` like so:
```python
PLUGINS = ['netbox_virtual_circuit_plugin'] # Note that the name here use underscore, not hyphen.
```

Don't forget to restart NetBox to load the new plugin.

You might also have to manually run the database migrations for Netbox to create the appropriate tables for virtual circuits.
```bash
python3 manage.py migrate
```

For more information about installing plugins, refer to [NetBox's documentation.](https://netbox.readthedocs.io/en/stable/plugins/)

## Compatibility

Below is a table describing the compatibility of various NetBox versions with NetBox Virtual Circuit Plugin versions.

|                   | Netbox 2.8 | NetBox 2.9 |
| ----------------- | ---------- | ---------- |
| `1.0.x`           | ✓          | ✗          |
| `1.1.x` and above | ✗          | ✓          |

## Using

Once the plugin is installed correctly as instructed above, one can find the
**Virtual Circuit** section under **Plugins** navigation tab via NetBox UI
that is ready to use with correct *admin* permission.

As for REST API use cases, the 2 group endpoints are exposed at:
- `/api/plugins/virtual-circuit/virtual-circuits`
- `/api/plugins/virtual-circuit/vlans`

While the former one is for creating/retrieving/modifying/deleting Virtual
Circuits, the later one is for assigning and managing Virtual-Circuit-to-VLAN
connections. For more information, refer to `/api/docs` as it also conforms
to Swagger Specification for hosted visual documentations.

## Developing

Plugins are essentially self-contained Django apps which integrate with NetBox
to provide custom functionality. For more information, see [NetBox
documentation](https://netbox.readthedocs.io/en/stable/plugins/development/).

To help setup the development environment, it comes with a CLI helper
based on `Makefile`, including the following commands:
```
changelog        Generate a changelog file from GitHub Issue Tracker
clean            Clean up build artifacts
deploy           Run a local development deployment of the plugin with NetBox
docker           Build a local docker image
github-tag       Create and push a tag with the current version
help             Print usage information
migrate          Run makemigrations in Django and produce a migration file locally
release          Package and distribute the current release to PyPI
test             Run unit tests
version          Print the version
```

That said, one can simply build and run a local development image of the plugin
with NetBox with a single line:
```
make build && make deploy
```

The application will be available after a few minutes at
`http://0.0.0.0:8000/`. The default credentials are:
- Username: **admin**
- Password: **admin**
- API Token: **0123456789abcdef0123456789abcdef01234567**

Below are several screenshots of the UI:

![Navigation view](docs/images/1_navigation.png)

![Add a Virtual Circuit](docs/images/2_add.png)

![List all Virtual Circuits](docs/images/3_list.png)

![Assign a VLAN to a Virtual Circuit](docs/images/4_assign_vlan.png)

![Virtual Circuit Single View (2 VLANs assigned)](docs/images/5_assigned_vlans.png)

![List of all connections](docs/images/6_connections.png)

## Contributing

If you experience a bug, would like to ask a question, or request a feature,
open a new issue and provide as much context as possible. All contributions,
questions, and feedback are welcomed and appreciated.

## License

NetBox Virtual Circuit Plugin is licensed under GPLv3. See [LICENSE](LICENSE)
for more info.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/HEAD)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v1.4.0...HEAD)

**Implemented enhancements:**

- `vlan\_of` related name is awkward [\#41](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/41)

## [v1.4.0](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v1.4.0) (2020-10-30)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v1.1.0...v1.4.0)

**Implemented enhancements:**

- Add a box for Virtual Circuits in the tenant information screen. [\#40](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/40)

**Closed issues:**

- Include device in VLAN list next to interface [\#42](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/42)
- netbox:: ChangeLoggedModel has moved location and is now in extras [\#37](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/37)

**Merged pull requests:**

- Update VCVLAN related name [\#46](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/46) ([hoanhan101](https://github.com/hoanhan101))
- Inject Virtual Circuit counts to Tenant screen [\#45](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/45) ([hoanhan101](https://github.com/hoanhan101))
- Include VLAN's interface parent [\#44](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/44) ([hoanhan101](https://github.com/hoanhan101))

## [v1.1.0](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v1.1.0) (2020-09-22)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v1.0.0...v1.1.0)

**Fixed bugs:**

- Plugin fails jsonschema/open api validation [\#34](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/34)

**Closed issues:**

- Setup GitHub Actions pipeline [\#30](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/30)

**Merged pull requests:**

- Fix NetBox 2.9.3 compatibility issues [\#38](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/38) ([hoanhan101](https://github.com/hoanhan101))

## [v1.0.0](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v1.0.0) (2020-08-21)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v0.3.1...v1.0.0)

**Implemented enhancements:**

- Add VCID plugin info inside "Change Log" [\#31](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/31)

**Merged pull requests:**

- Fix swagger missing id [\#35](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/35) ([hoanhan101](https://github.com/hoanhan101))
- Include change log [\#32](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/32) ([hoanhan101](https://github.com/hoanhan101))
- Set VCID's range from 1 to 4294967295 [\#28](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/28) ([hoanhan101](https://github.com/hoanhan101))

## [v0.3.1](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v0.3.1) (2020-07-30)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v0.3.0...v0.3.1)

## [v0.3.0](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v0.3.0) (2020-07-21)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v0.2.1...v0.3.0)

**Fixed bugs:**

- VCID modeling problem [\#23](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/23)

**Closed issues:**

- Add `Deleted` status [\#26](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/26)

**Merged pull requests:**

- Run makemigrations in Django and produce a migration file locally [\#29](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/29) ([hoanhan101](https://github.com/hoanhan101))
- Update status to better reflect the provisioning process [\#27](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/27) ([hoanhan101](https://github.com/hoanhan101))
- Include docs in source distribution [\#25](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/25) ([hoanhan101](https://github.com/hoanhan101))

## [v0.2.1](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v0.2.1) (2020-07-07)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/v0.2.0...v0.2.1)

**Closed issues:**

- Breaks Swagger API [\#22](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/22)

**Merged pull requests:**

- Add command for creating and pushing a tag with the current version [\#21](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/21) ([hoanhan101](https://github.com/hoanhan101))

## [v0.2.0](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/tree/v0.2.0) (2020-06-18)

[Full Changelog](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/compare/e795f6529d5e6f76c1d541f0bfae8c650f0cc7a7...v0.2.0)

**Closed issues:**

- Plugin netbox-virtual-circuit-plugin does not provide a 'config' variable. [\#8](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/issues/8)

**Merged pull requests:**

- Flesh out UI view [\#20](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/20) ([hoanhan101](https://github.com/hoanhan101))
- Check for existing Virtual-Circuit-to-VLAN connection before creating one [\#19](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/19) ([hoanhan101](https://github.com/hoanhan101))
- Expand `vlans` field inside a virtual circuit object response [\#15](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/15) ([hoanhan101](https://github.com/hoanhan101))
- Add unit tests [\#14](https://github.com/vapor-ware/netbox-virtual-circuit-plugin/pull/14) ([hoanhan101](https://github.com/hoanhan101))

netbox_virtual_circuit_plugin/models.py
from django.db import models
from django.core.validators import MaxValueValidator, MinValueValidator
from django.urls import reverse
from ipam.models import VLAN
from extras.models import ChangeLoggedModel

from .choices import VirtualCircuitStatusChoices


class VirtualCircuit(ChangeLoggedModel):
    """Virtual Circuit model."""

    vcid = models.BigIntegerField(
        primary_key=True,
        verbose_name='ID',
        validators=[
            MaxValueValidator(4294967295),
            MinValueValidator(1),
        ],
    )
    name = models.CharField(
        max_length=64,
    )
    status = models.CharField(
        max_length=30,
        choices=VirtualCircuitStatusChoices,
        default=VirtualCircuitStatusChoices.STATUS_PENDING_CONFIGURATION,
    )
    context = models.CharField(
        max_length=100,
        blank=True,
    )

    class Meta:
        ordering = ['vcid']
        verbose_name = 'Virtual Circuit'
        verbose_name_plural = 'Virtual Circuits'

    def __str__(self):
        return f'{self.vcid} ({self.name})'

    def get_absolute_url(self):
        return reverse('plugins:netbox_virtual_circuit_plugin:virtual_circuit', args=[self.vcid])

class VirtualCircuitVLAN(ChangeLoggedModel):
    """Virtual Circuit to VLAN relationship."""

    virtual_circuit = models.ForeignKey(
        to=VirtualCircuit,
        on_delete=models.CASCADE,
        related_name='vlans',
        verbose_name='Virtual Circuit',
    )
    vlan = models.OneToOneField(
        to=VLAN,
        on_delete=models.CASCADE,
        related_name='virtual_circuit',
        verbose_name='VLAN',
    )

    class Meta:
        ordering = ['virtual_circuit']
        verbose_name = 'Virtual-Circuit-to-VLAN connection'
        verbose_name_plural = 'Virtual-Circuit-to-VLAN connections'

    def get_absolute_url(self):
        return reverse('plugins:netbox_virtual_circuit_plugin:virtual_circuit', args=[self.virtual_circuit.vcid])
